File: 1.two-sum.cs
/*
 * @lc app=leetcode id=1 lang=csharp
 *
 * [1] Two Sum
 */

// @lc code=start
using System;
using System.Collections.Generic;

public class Solution {
    public int[] TwoSum(int[] nums, int target)
    {
        //         <Key,Value>
        Dictionary<int, int> map = new Dictionary<int, int>(); //使用Dictionary設置Hash Table

        for(int i=0; i<nums.Length; i++)                       //造訪串列內所有數字
        {
            int complement = target - nums[i];                 //總和的數字 減掉當前的數字後，將剩餘的數字放入complement
            if(map.ContainsKey(complement))                    //在Hash Table找找看有沒有符合complement的數字
            {
                return new int[] {map[complement], i};         //如果有符合，就回傳答案
            }

            map[nums[i]] = i;   //不符合的話，把目前的數字nums[i]作為key，index i作為Value暫存起來
        }
        return new int[0];
    }
}
// @lc code=end

File: 1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start
from typing import List
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        #使用HashMap
        map = {} #key : value
        for i in range(len(nums)):
            #總和的數字，減掉當前的數字，將剩餘的數字放入complement
            complement = target - nums[i]
            if (complement in map):
                return [map[complement], i] #Map有符合的complement，就回傳該complement的索引值和當前的i索引值
            map[nums[i]] = i #如果Map沒有符合的complement，就把當前數字和數字的索引值分別以key和value放進Map內
        return
# @lc code=end

